pytest_jupyterhub/jupyterhub_spawners.py

```diff
@@ -25,3 +29,29 @@ def jupyterhub_spawner_class_config(spawner_class):
 def spawner_config():
     """Allows tests to setup configurations that are specific to the spawner implementation"""
     return Config()
+
+
+@pytest.fixture
+def app_for_spawners(request, jupyterhub_spawner_class_config, spawner_config):
+    """Creates an instance of a JupyterHub application that can be used by different spawner implementation test suites"""
+    final_config = jupyterhub_spawner_class_config
+    final_config.merge(spawner_config)
+    mocked_app = MockHub.instance(config=final_config)
+
+    async def make_app():
+        """Initializes the mocked application and starts it"""
+        await mocked_app.initialize([])
+        await mocked_app.start()
+
+    def fin():
+        """Disconnects logging during cleanup because pytest closes captured FDs prematurely"""
+        mocked_app.log.handlers = []
+        MockHub.clear_instance()
+        try:
+            mocked_app.stop()
+        except Exception as e:
+            print("Error stopping Hub: %s" % e, file=sys.stderr)
+
+    request.addfinalizer(fin)
+    asyncio.run(make_app())
```

Comment:
I actually think we should also be bringing the `io_loop` fixture from the `jupyterhub` repo here and use it to start the hub app.

https://github.com/jupyterhub/jupyterhub/blob/95cf06a46e2290ddc5fe9a8b0b2152eebfd2253d/jupyterhub/tests/conftest.py#L133

But I don't feel comfortable and knowledgeable enough to explain how the `tornado` and `asyncio` libraries interact and integrate with each other in the same event loop in `jupyterhub` and its test suite.

@minrk, would you be willing to provide us with some context about this please? it can be either async in or by joining Sheila's and mine's meeting tomorrow?